kernel/src/fs/utils/range_lock/mod.rs

```diff
@@ -302,10 +317,10 @@ impl RangeLockList {
                     next_idx += 1;
                 } else {
                     // Split overlapping locks
-                    let overlap_with = pre_lock.overlap_with(&next_lock).unwrap();
+                    let overlap_with = pre_lock.overlap_with(next_lock).unwrap();
                     match overlap_with {
                         OverlapWith::ToLeft => {
-                            list[next_idx].set_start(pre_lock.end());
+                            next_lock.set_start(pre_lock.end());
```

Comment:
Not introduced in this PR, so this is mostly a question: Won't the algorithm screw up the waiters in the queue? I can hardly see why this algorithm is correct with respect to the waiters in the queue.

In other words, if the start and end positions can be updated, how to guarantee that the waiters end up in the right queue?

Reply:
If the range shrinks, it will wake up all waiters. These waiters will then `try_set_lock()` in `pause_until()`. If a conflict lock still exists, they'll re-enqueue a waker in the conflict lock's waitqueue.

https://github.com/asterinas/asterinas/blob/f93ba70c4b794742e415869fb35ff9c9fe057e01/kernel/src/fs/utils/range_lock/mod.rs#L114-L123

Reply:
Thanks for the explanation! This makes sense to me.

For the commented code, two locks of _different types_ are actually overlapping, but how would that ever be possible? I think read locks and write locks cannot overlap.

Also, the start of `next_lock` is updated, but `next_lock` and `pre_lock` have different types, so the type for the overlapping region is changed. Is this expected?

Reply:
Yes, it is expected (taken from TLPI book).
![tlpi](https://github.com/user-attachments/assets/bd05dd1b-a65e-4df4-9d63-e3a1bd5bc39e)

Reply:
Good to know! Thanks for the explanation!